03-rnaseq/gene-id-annotation_rnaseq_01_ensembl.Rmd

```diff
@@ -0,0 +1,380 @@
+We have a total of 235 out of 31882 Ensembl IDs with multiple mappings to Entrez IDs.
+With that rate, we can filter out the multi-mapped IDs if we want to keep 1:1 mappings for our downstream analyses by supplying the `"filter"` option to the `multiVals` argument of `mapIds()`.
+If you want to see an example of a `multiVals = "filter"` strategy see the [microarray gene identifier annotation example notebook](https://github.com/AlexsLemonade/refinebio-examples/tree/master/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+### Store all mapped values
+
+We are not sure which Entrez ID is relevant as yet, but we could store all of the mapped information for future use.
+In the next chunk, we show how we can collapse all the Entrez IDs into one column where we store them all for each unique Ensembl ID.
+
+```{r}
+final_mapped_df <- df %>%
+  # Collapse the mapped values in `mapped_list` into one column named `entrez_ids`
+  dplyr::mutate(entrez_ids = sapply(mapped_list, paste, collapse = ";")) %>%
+  # Select the `entrez_ids` column first, then everything else in the data frame
+  dplyr::select(entrez_ids, dplyr::everything()) %>%
+  # Now arrange by the new `entrez_ids` column
+  dplyr::arrange(entrez_ids)
+```
+
+Let's take a look at our new collapsed `entrez_ids` column in the `final_mapped_df` object.
+
+```{r}
+final_mapped_df %>%
+  # Filter `final_mapped_df` to include only the rows where `entrez_ids` values include the ";" character -- we know that these are the rows with multiple mapped values
+  dplyr::filter(stringr::str_detect(entrez_ids, ";")) %>%
+  # We only need a preview here
+  head()
+```
+
+You may have a list of Entrez IDs you are interested in, in which case, the above format may work best for you.
+In a different study, you may want the oldest Entrez ID (which is probably first), in which case, you can create a column that stores just the first mapped Entrez ID that comes up for each Ensembl ID.
+We will do this in the next section by re-running the `mapIds()` function with `multiVals = "first"`.
+
+### Map Ensembl IDs to gene symbols -- keeping only the first mapped value
+
+Let's re-run `mapIds()`, this time using `multiVals = "first"` and add this information to our `final_mapped_df` object.
+
+```{r}
+final_mapped_df <- data.frame(
+  "first_mapped_entrez_id" = mapIds(
+    org.Dr.eg.db, # Replace with annotation package for the organism relevant to your data
+    keys = df$Gene,
+    column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
+    keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
+    multiVals = "first" # This will keep only the first mapped value for each Ensembl ID
+  )
+) %>%
+  # Make an `Ensembl` column to store the rownames
+  tibble::rownames_to_column("Ensembl") %>%
+  # Join the remaining data from `df` using the Ensembl IDs
+  dplyr::inner_join(final_mapped_df, by = c("Ensembl" = "Gene"))
```

Comment:
Just for consistency, since the other columns will have `id` in the name.
```suggestion
  tibble::rownames_to_column("Ensembl_id") %>%
  # Join the remaining data from `df` using the Ensembl IDs
  dplyr::inner_join(final_mapped_df, by = c("Ensembl_id" = "Gene"))
```

Reply:
I maybe don't like this after seeing how the data comes out of mapIds. Ignore this and stick with just plain `Ensembl`